modules/custom/dkan_api/src/Controller/Dataset.php

```diff
@@ -2,24 +2,49 @@
 
 namespace Drupal\dkan_api\Controller;
 
-use Drupal\dkan_api\Storage\DrupalNodeDataset;
-use Drupal\dkan_schema\SchemaRetriever;
-use JsonSchemaProvider\Provider;
+use Symfony\Component\DependencyInjection\ContainerInterface;
 
+/**
+ * Dataset.
+ */
 class Dataset extends Api {
 
-  private $nodeDataset;
+  use \Drupal\dkan_common\Util\TestableParentCallTrait;
 
-  public function __construct() {
-    $this->nodeDataset = new DrupalNodeDataset();
+  /**
+   *
+   * @var \Drupal\dkan_api\Storage\DrupalNodeDataset
+   */
+  protected $nodeDataset;
+
+  /**
+   * {@inheritdocs}.
+   */
+  public function __construct(ContainerInterface $container) {
```

Comment:
I believe route controllers are a special thing, but doesn't hurt to ask. Don't we have a way to inject specific services instead of the whole container?

Reply:
Not sure what you mean by special thing. Drupal has a `ContainerAwareInterface` that can be appled to anything really to make passing the service container easier. It's not specific to controllers.

it's a reference to the global service container. it won't affect performance or memory. It's simpler to pass the service container and use the `get` method as needed to get other services.

Reply:
@ketwaroo the question is, can you choose not to pass the container, and request specific services instead?

Reply:
Yes. see the static `::create()` method. Drupal uses that as factory for the container aware classes. 

I pass the only service container to save time/effort right now. Otherwise it would take me longer to comb through the whole class, figure out what can be converted to a service for injection, refactor everything, make sure it still works, etc.